// @datastax/astra-db-ts/dist/administration/data-api-db-admin.js
"use strict";
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
// noinspection ExceptionCaughtLocallyJS
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _DataAPIDbAdmin_httpClient, _DataAPIDbAdmin_db;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAPIDbAdmin = void 0;
const db_admin_1 = require("../administration/db-admin");
const spawn_admin_1 = require("../client/parsers/spawn-admin");
const constants_1 = require("../lib/constants");
class DataAPIDbAdmin extends db_admin_1.DbAdmin {
        constructor(db, httpClient, rawAdminOpts) {
        super();
        _DataAPIDbAdmin_httpClient.set(this, void 0);
        _DataAPIDbAdmin_db.set(this, void 0);
        const adminOpts = (0, spawn_admin_1.parseAdminSpawnOpts)(rawAdminOpts, 'options');
        __classPrivateFieldSet(this, _DataAPIDbAdmin_httpClient, httpClient.forDbAdmin(adminOpts), "f");
        __classPrivateFieldSet(this, _DataAPIDbAdmin_db, db, "f");
        Object.defineProperty(this, constants_1.$CustomInspect, {
            value: () => `DataAPIDbAdmin()`,
        });
    }
        db() {
        return __classPrivateFieldGet(this, _DataAPIDbAdmin_db, "f");
    }
        async findEmbeddingProviders(options) {
        const resp = await __classPrivateFieldGet(this, _DataAPIDbAdmin_httpClient, "f").executeCommand({ findEmbeddingProviders: {} }, {
            timeoutManager: __classPrivateFieldGet(this, _DataAPIDbAdmin_httpClient, "f").tm.single('databaseAdminTimeoutMs', options),
            methodName: 'dbAdmin.findEmbeddingProviders',
            keyspace: null,
        });
        return resp.status;
    }
        async listKeyspaces(options) {
        const resp = await __classPrivateFieldGet(this, _DataAPIDbAdmin_httpClient, "f").executeCommand({ findKeyspaces: {} }, {
            timeoutManager: __classPrivateFieldGet(this, _DataAPIDbAdmin_httpClient, "f").tm.single('keyspaceAdminTimeoutMs', options),
            methodName: 'dbAdmin.listKeyspaces',
            keyspace: null,
        });
        return resp.status.keyspaces;
    }
        async createKeyspace(keyspace, options) {
        if (options?.updateDbKeyspace) {
            __classPrivateFieldGet(this, _DataAPIDbAdmin_db, "f").useKeyspace(keyspace);
        }
        const replication = options?.replication ?? {
            class: 'SimpleStrategy',
            replicationFactor: 1,
        };
        await __classPrivateFieldGet(this, _DataAPIDbAdmin_httpClient, "f").executeCommand({ createKeyspace: { name: keyspace, options: { replication } } }, {
            timeoutManager: __classPrivateFieldGet(this, _DataAPIDbAdmin_httpClient, "f").tm.single('keyspaceAdminTimeoutMs', options),
            methodName: 'dbAdmin.createKeyspace',
            keyspace: null,
        });
    }
        async dropKeyspace(keyspace, options) {
        await __classPrivateFieldGet(this, _DataAPIDbAdmin_httpClient, "f").executeCommand({ dropKeyspace: { name: keyspace } }, {
            timeoutManager: __classPrivateFieldGet(this, _DataAPIDbAdmin_httpClient, "f").tm.single('keyspaceAdminTimeoutMs', options),
            methodName: 'dbAdmin.dropKeyspace',
            keyspace: null,
        });
    }
    get _httpClient() {
        return __classPrivateFieldGet(this, _DataAPIDbAdmin_httpClient, "f");
    }
}
exports.DataAPIDbAdmin = DataAPIDbAdmin;
_DataAPIDbAdmin_httpClient = new WeakMap(), _DataAPIDbAdmin_db = new WeakMap();
