// @datastax/astra-db-ts/dist/administration/types/admin/database-info.js
"use strict";
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
// // export interface RawAstraDbAdminInfo {
//   //   id: string,
//   //   orgId: string,
//   //   ownerId: string,
//   //   info: RawAstraDbInfo,
//   //   creationTime?: string,
//   //   lastUsageTime?: string,
//   //   terminationTime?: string,
//   //   status: AstraDbStatus,
//   //   observedStatus: AstraDbStatus,
//   //   storage?: AstraDbStorageInfo,
//   //   availableActions?: AstraDbAction[],
//   //   cost?: AstraDbCostInfo,
//   //   message?: string,
//   //   grafanaUrl?: string,
//   //   cqlshUrl?: string,
//   //   graphqlUrl?: string,
//   //   dataEndpointUrl?: string,
//   //   metrics?: AstraDbMetricsInfo,
// }
//
// // export interface RawAstraDbInfo {
//   //   name: string,
//   //   keyspace?: string,
//   //   cloudProvider?: AstraDbCloudProvider,
//   //   tier: AstraDbTier,
//   //   capacityUnits: number,
//   //   region: string,
//   //   user?: string,
//   //   password?: string,
//   //   additionalKeyspaces?: string[],
//   //   keyspaces?: string[],
//   //   dbType?: 'vector',
//   //   datacenters?: AstraDbDatacenterInfo[],
// }
//
// // export interface AstraDbStorageInfo {
//   //   nodeCount: number,
//   //   replicationFactor: number,
//   //   totalStorage: number,
//   //   usedStorage?: number,
// }
//
// // export interface AstraDbDatacenterInfo {
//   //   capacityUnits: number,
//   //   cloudProvider: AstraDbCloudProvider,
//   //   dateCreated: string,
//   //   id: string,
//   //   isPrimary: boolean,
//   //   name: string,
//   //   region: string,
//   //   regionClassification: string,
//   //   regionZone: string,
//   //   secureBundleUrl: string,
//   //   status: string,
//   //   tier: AstraDbTier,
// }
//
// // export interface AstraDbCostInfo {
//   //   costPerDayCents: number,
//   //   costPerDayMRCents: number,
//   //   costPerDayParkedCents: number,
//   //   costPerHourCents: number,
//   //   costPerHourMRCents: number,
//   //   costPerHourParkedCents: number,
//   //   costPerMinCents: number,
//   //   costPerMinMRCents: number,
//   //   costPerMinParkedCents: number,
//   //   costPerMonthCents: number,
//   //   costPerMonthMRCents: number,
//   //   costPerMonthParkedCents: number,
//   //   costPerNetworkGbCents: number,
//   //   costPerReadGbCents: number,
//   //   costPerWrittenGbCents: number,
// }
//
// // export interface AstraDbMetricsInfo {
//   //   errorsTotalCount: number,
//   //   liveDataSizeBytes: number,
//   //   readRequestsTotalCount: number,
//   //   writeRequestsTotalCount: number
// }
