// @datastax/astra-db-ts/dist/documents/tables/table.js
"use strict";
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _Table_httpClient, _Table_commands, _Table_db;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Table = void 0;
const documents_1 = require("../../documents");
const command_impls_1 = require("../../documents/commands/command-impls");
const constants_1 = require("../../lib/constants");
const json_bigint_1 = __importDefault(require("json-bigint"));
const cursor_1 = require("../../documents/tables/cursor");
const utils_1 = require("../../lib/utils");
const ser_des_1 = require("../../documents/tables/ser-des/ser-des");
const jbi = (0, json_bigint_1.default)({ storeAsString: true });
class Table {
        constructor(db, httpClient, name, opts) {
        _Table_httpClient.set(this, void 0);
        _Table_commands.set(this, void 0);
        _Table_db.set(this, void 0);
                Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
                Object.defineProperty(this, "keyspace", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, 'name', {
            value: name,
        });
        Object.defineProperty(this, 'keyspace', {
            value: opts?.keyspace ?? db.keyspace,
        });
        const hack = {
            parseWithBigNumbers(json) {
                return json.includes('{"type":"varint"}') || json.includes('{"type":"decimal"}');
            },
            parser: (0, utils_1.withJbiNullProtoFix)(jbi),
        };
        __classPrivateFieldSet(this, _Table_httpClient, httpClient.forTableSlashCollectionOrWhateverWeWouldCallTheUnionOfTheseTypes(this.keyspace, this.name, opts, hack), "f");
        __classPrivateFieldSet(this, _Table_commands, new command_impls_1.CommandImpls(this, __classPrivateFieldGet(this, _Table_httpClient, "f"), new ser_des_1.TableSerDes(opts?.serdes)), "f");
        __classPrivateFieldSet(this, _Table_db, db, "f");
        Object.defineProperty(this, constants_1.$CustomInspect, {
            value: () => `Table(keyspace="${this.keyspace}",name="${this.name}")`,
        });
    }
        async insertOne(row, timeout) {
        return __classPrivateFieldGet(this, _Table_commands, "f").insertOne(row, timeout);
    }
        async insertMany(rows, options) {
        return __classPrivateFieldGet(this, _Table_commands, "f").insertMany(rows, options, documents_1.TableInsertManyError);
    }
        async updateOne(filter, update, timeout) {
        await __classPrivateFieldGet(this, _Table_commands, "f").updateOne(filter, update, timeout);
    }
        async deleteOne(filter, timeout) {
        await __classPrivateFieldGet(this, _Table_commands, "f").deleteOne(filter, timeout);
    }
        async deleteMany(filter, timeout) {
        await __classPrivateFieldGet(this, _Table_commands, "f").deleteMany(filter, timeout);
    }
    find(filter, options) {
        return __classPrivateFieldGet(this, _Table_commands, "f").find(filter, options, cursor_1.TableFindCursor);
    }
    async findOne(filter, options) {
        return __classPrivateFieldGet(this, _Table_commands, "f").findOne(filter, options);
    }
        async alter(options) {
        await __classPrivateFieldGet(this, _Table_httpClient, "f").executeCommand({
            alterTable: {
                operation: options.operation,
            },
        }, {
            timeoutManager: __classPrivateFieldGet(this, _Table_httpClient, "f").tm.single('tableAdminTimeoutMs', options),
        });
        return this;
    }
    async listIndexes(options) {
        const resp = await __classPrivateFieldGet(this, _Table_httpClient, "f").executeCommand({
            listIndexes: {
                options: {
                    explain: options?.nameOnly !== true,
                },
            },
        }, {
            timeoutManager: __classPrivateFieldGet(this, _Table_httpClient, "f").tm.single('tableAdminTimeoutMs', options),
        });
        return resp.status.indexes;
    }
        async createIndex(name, column, options) {
        await __classPrivateFieldGet(this, _Table_httpClient, "f").executeCommand({
            createIndex: {
                name: name,
                definition: {
                    column,
                    options: {
                        caseSensitive: options?.options?.caseSensitive,
                        normalize: options?.options?.normalize,
                        ascii: options?.options?.ascii,
                    },
                },
                options: {
                    ifNotExists: options?.ifNotExists,
                },
            },
        }, {
            timeoutManager: __classPrivateFieldGet(this, _Table_httpClient, "f").tm.single('tableAdminTimeoutMs', options),
        });
    }
        async createVectorIndex(name, column, options) {
        await __classPrivateFieldGet(this, _Table_httpClient, "f").executeCommand({
            createVectorIndex: {
                name: name,
                definition: {
                    column,
                    options: {
                        sourceModel: options?.options?.sourceModel,
                        metric: options?.options?.metric,
                    },
                },
                options: {
                    ifNotExists: options?.ifNotExists,
                },
            },
        }, {
            timeoutManager: __classPrivateFieldGet(this, _Table_httpClient, "f").tm.single('tableAdminTimeoutMs', options),
        });
    }
        async definition(options) {
        const resp = await __classPrivateFieldGet(this, _Table_db, "f").listTables({
            timeout: options?.timeout,
            keyspace: this.keyspace,
        });
        const table = resp.find((t) => t.name === this.name);
        if (!table) {
            throw new Error(`Can not get definition for table '${this.keyspace}.${this.name}'; table not found. Did you use the right keyspace?`);
        }
        return table.definition;
    }
        async drop(options) {
        await __classPrivateFieldGet(this, _Table_db, "f").dropTable(this.name, { ...options, keyspace: this.keyspace });
    }
        get _httpClient() {
        return __classPrivateFieldGet(this, _Table_httpClient, "f");
    }
}
exports.Table = Table;
_Table_httpClient = new WeakMap(), _Table_commands = new WeakMap(), _Table_db = new WeakMap();
