// @datastax/astra-db-ts/dist/documents/index.js
"use strict";
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CollectionUpdateManyError = exports.TooManyRowsToCountError = exports.TooManyDocumentsToCountError = exports.TableInsertManyError = exports.CollectionInsertManyError = exports.CollectionDeleteManyError = exports.DataAPIError = exports.CumulativeOperationError = exports.DataAPITimeoutError = exports.DataAPIHttpError = exports.DataAPIResponseError = void 0;
__exportStar(require("./cursor"), exports);
__exportStar(require("./events"), exports);
__exportStar(require("./datatypes/object-id"), exports);
__exportStar(require("./collections"), exports);
__exportStar(require("./tables"), exports);
__exportStar(require("./embedding-providers"), exports);
__exportStar(require("./datatypes"), exports);
var errors_1 = require("./errors");
Object.defineProperty(exports, "DataAPIResponseError", { enumerable: true, get: function () { return errors_1.DataAPIResponseError; } });
Object.defineProperty(exports, "DataAPIHttpError", { enumerable: true, get: function () { return errors_1.DataAPIHttpError; } });
Object.defineProperty(exports, "DataAPITimeoutError", { enumerable: true, get: function () { return errors_1.DataAPITimeoutError; } });
Object.defineProperty(exports, "CumulativeOperationError", { enumerable: true, get: function () { return errors_1.CumulativeOperationError; } });
Object.defineProperty(exports, "DataAPIError", { enumerable: true, get: function () { return errors_1.DataAPIError; } });
Object.defineProperty(exports, "CollectionDeleteManyError", { enumerable: true, get: function () { return errors_1.CollectionDeleteManyError; } });
Object.defineProperty(exports, "CollectionInsertManyError", { enumerable: true, get: function () { return errors_1.CollectionInsertManyError; } });
Object.defineProperty(exports, "TableInsertManyError", { enumerable: true, get: function () { return errors_1.TableInsertManyError; } });
Object.defineProperty(exports, "TooManyDocumentsToCountError", { enumerable: true, get: function () { return errors_1.TooManyDocumentsToCountError; } });
Object.defineProperty(exports, "TooManyRowsToCountError", { enumerable: true, get: function () { return errors_1.TooManyRowsToCountError; } });
Object.defineProperty(exports, "CollectionUpdateManyError", { enumerable: true, get: function () { return errors_1.CollectionUpdateManyError; } });
